transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,598 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Background task dedicated to manage the QUIC state machine.
+//!
+//! Considering that all QUIC communications happen over a single UDP socket, one needs to
+//! maintain a unique synchronization point that holds the state of all the active connections.
+//!
+//! The endpoint represents this synchronization point. It is maintained in a background task
+//! whose role is to interface with the UDP socket. Communication between the background task and
+//! the rest of the code only happens through channels. See the documentation of the
+//! [`EndpointDriver`] for a thorough description.
+
+use crate::{
+    connection::Connection,
+    tls,
+    transport::{self, Provider},
+};
+
+use futures::{
+    channel::{mpsc, oneshot},
+    prelude::*,
+    ready,
+};
+use std::{
+    collections::HashMap,
+    io,
+    net::{Ipv4Addr, Ipv6Addr, SocketAddr},
+    ops::ControlFlow,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+use x509_parser::nom::AsBytes;
+
+/// Represents the configuration for the QUIC endpoint.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The client configuration to pass to `quinn_proto`.
+    client_config: quinn_proto::ClientConfig,
+    /// The server configuration to pass to `quinn_proto`.
+    server_config: Arc<quinn_proto::ServerConfig>,
+    /// The endpoint configuration to pass to `quinn_proto`.
+    endpoint_config: Arc<quinn_proto::EndpointConfig>,
+}
+
+impl Config {
+    /// Creates a new configuration object with default values.
+    pub fn new(keypair: &libp2p_core::identity::Keypair) -> Result<Self, tls::ConfigError> {
+        let mut transport = quinn_proto::TransportConfig::default();
+        transport.max_concurrent_uni_streams(0u32.into()); // Can only panic if value is out of range.
```

Comment:
What's the bidirectional stream limit?

Reply:
100

Reply:
I assume that this is only counting peer-initiated streams, right?
100 seems a bit low, we had people complain when we lowered the limit. go-libp2p is setting it to 256 now.

Reply:
Included 256 as default in 2264ab83377758d0644e668caba1ae9048cc384c (but configurable by the user).